File: 739.每日温度.java
/*
 * @lc app=leetcode.cn id=739 lang=java
 *
 * [739] 每日温度
 */

// @lc code=start


class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] res = new int[n];

        Stack<Integer> stack = new Stack<Integer>();

        // 单调栈，从后往前找，把比当前元素小的都删掉
        // 栈里存大元素的下标
        for (int i = n - 1; i >= 0; i --) {
            while(!stack.isEmpty() && temperatures[stack.peek()] <= temperatures[i]) {
                stack.pop();
            }
            // 如果栈为空，则当前数最大，返回值为0，
            //  如果栈不为空，则用栈顶存的下标减去当前下标即为结果
            res[i] = stack.isEmpty() ? 0 : (stack.peek() - i);
            stack.push(i);
        }

        return res;
    }
}
// @lc code=end

File: 739.每日温度.js
/*
 * @lc app=leetcode.cn id=739 lang=javascript
 *
 * [739] 每日温度
 */

// @lc code=start
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    
    let n = temperatures.length;
    let stack = [];
    let res = Array(n);

    for (let i = n - 1; i >= 0; i--) {
        while (stack.length && temperatures[stack[stack.length - 1]] <= temperatures[i]) {
            stack.pop();
        }
        if (stack.length) {
            res[i] = stack[stack.length - 1] - i;
        } else {
            res[i] = 0;
        }
        stack.push(i);
    }

    return res;

};
// @lc code=end

